homeassistant/components/tado/sensor.py

```diff
@@ -149,97 +144,138 @@ def icon(self):
 
         elif self.zone_variable == "overlay":
-            self._state = "overlay" in data and data["overlay"] is not None
+            self._state = self._tado_zone_data.overlay_active
             self._state_attributes = (
-                {"termination": data["overlay"]["termination"]["type"]}
-                if self._state
+                {"termination": self._tado_zone_data.overlay_termination_type}
+                if self._tado_zone_data.overlay_active
                 else {}
             )
 
         elif self.zone_variable == "early start":
-            self._state = "preparation" in data and data["preparation"] is not None
+            self._state = self._tado_zone_data.preparation
 
         elif self.zone_variable == "open window":
-            self._state = "openWindow" in data and data["openWindow"] is not None
-            self._state_attributes = data["openWindow"] if self._state else {}
+            self._state = self._tado_zone_data.open_window
+            self._state_attributes = self._tado_zone_data.open_window_attr
+
+
+class TadoDeviceSensor(Entity):
+    """Representation of a tado Sensor."""
+
+    def __init__(self, hass, tado, device_name, device_id, device_variable):
+        """Initialize of the Tado Sensor."""
+        self.hass = hass
+        self._tado = tado
+
+        self.device_name = device_name
+        self.device_id = device_id
+        self.device_variable = device_variable
+
+        self._unique_id = f"{device_variable} {device_id} {tado.device_id}"
+
+        self._state = None
+        self._state_attributes = None
+        self._tado_device_data = None
+        self._async_update_device_data()
+
+    async def async_added_to_hass(self):
+        """Register for sensor updates."""
+
+        async_dispatcher_connect(
+            self.hass,
+            SIGNAL_TADO_UPDATE_RECEIVED.format("device", self.device_id),
+            self._async_update_callback,
+        )
+
+    @property
+    def unique_id(self):
+        """Return the unique id."""
+        return self._unique_id
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return f"{self.device_name} {self.device_variable}"
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def should_poll(self):
+        """Do not poll."""
+        return False
+
+    @callback
+    def _async_update_callback(self):
+        """Update and write state."""
+        self._async_update_device_data()
+        self.async_write_ha_state()
+
+    @callback
+    def _async_update_device_data(self):
+        """Handle update callbacks."""
+        try:
+            data = self._tado.data["device"][self.device_id]
+        except KeyError:
+            return
+
+        if self.device_variable == "tado bridge status":
+            self._state = data.get("connectionState", {}).get("value", False)
```

Comment:
This looks like it should be a binary sensor ?

Reply:
Yes, this is existing and changing it would be a breaking change if anyone is using it.

Reply:
In a future PR, let's add the binary sensor one and we can keep the sensor one around for a bit, but disable by default so new integrations won't end up using it.

Reply:
will do